### PR TITLE
Add meta parameter

### PR DESCRIPTION
**Pull request by robertholmes3**
Adding meta columns for sources and exposures. 

[dbt docs - meta](https://docs.getdbt.com/reference/resource-configs/meta)

### CODE REVIEW

File: models/sources/sources.sql [hunk 1]
```diff
@@ -10,14 +10,19 @@ select
...
+    cast(null as {{ type_string() }}) as source_description,
     cast(null as {{ type_string() }}) as loader,
     cast(null as {{ type_string() }}) as name,
     cast(null as {{ type_string() }}) as identifier,
     cast(null as {{ type_string() }}) as loaded_at_field,
     {% if target.type == 'snowflake'%}
-        cast(null as {{ type_array() }}) as freshness
+        cast(null as {{ type_array() }}) as freshness,
+        cast(null as {{ type_array() }}) as source_meta,
+        cast(null as {{ type_array() }}) as meta
```
Comment (NiallRees) [inline models/sources/sources.sql]:
```suggestion
```

File: models/sources/sources.sql [hunk 2]
```diff
@@ -10,14 +10,19 @@ select
...
     {% else %}
-        cast(null as {{ type_json() }}) as freshness
+        cast(null as {{ type_json() }}) as freshness,
+        cast(null as {{ type_json() }}) as source_meta,
+        cast(null as {{ type_json() }}) as meta
```
Comment (NiallRees) [inline models/sources/sources.sql:24]:
```suggestion
```
Please could you move `source_meta` and `meta` below the `endif`? They are always json types.

File: macros/upload_sources.sql [hunk 3]
```diff
@@ -57,11 +63,14 @@
...
                     '{{ source.source_name }}', {# source_name #}
+                    '{{ source.source_description }}', {# source_description #}
                     '{{ source.loader }}', {# loader #}
                     '{{ source.name }}', {# name #}
                     '{{ source.identifier }}', {# identifier #}
                     '{{ source.loaded_at_field | replace("'","\\'") }}', {# loaded_at_field #}
-                    parse_json('{{ tojson(source.freshness) | replace("'","\\'") }}')  {# freshness #}
+                    parse_json('{{ tojson(source.freshness) | replace("'","\\'") }}'),  {# freshness #}
+                    parse_json('{{ tojson(source.source_meta) | replace("'","\\'") }}'),  {# source_meta #}
+                    parse_json('{{ tojson(source.meta) | replace("'","\\'") }}')  {# meta #}
```
Comment (NiallRees) [inline macros/upload_sources.sql:73]: Please can you add examples of these configs to the sources.yml for the integration tests? https://github.com/brooklyn-data/dbt_artifacts/blob/main/integration_test_project/models/sources.yml

File: macros/upload_exposures.sql [hunk 4]
```diff
@@ -58,13 +62,15 @@
                     '{{ run_started_at }}', {# run_started_at #}
                     '{{ exposure.name | replace("'","\\'") }}', {# name #}
                     '{{ exposure.type }}', {# type #}
-                    parse_json('{{ tojson(exposure.owner) | replace("'","\\'") }}'), {# owner #}
+                    '{{ tojson(exposure.owner) | replace("'","\\'") }}', {# owner #}
```
Comment (NiallRees) [inline macros/upload_exposures.sql]: What's the thinking behind removing `parse_json`?

Comment (NiallRees) [reply macros/upload_exposures.sql]: This results in the following error:
```
Database Error
  Value has type STRING which cannot be inserted into column owner, which has type JSON at [15:21]
```

File: macros/upload_exposures.sql [hunk 5]
```diff
@@ -58,13 +62,15 @@
...
                     '{{ exposure.maturity }}', {# maturity #}
                     '{{ exposure.original_file_path | replace('\\', '\\\\') }}', {# path #}
-                    """{{ exposure.description | replace("'","\\'") }}""", {# description #}
+                    '{{ exposure.description | replace("'","\\'") }}', {# description #}
                     '{{ exposure.url }}', {# url #}
                     '{{ exposure.package_name }}', {# package_name #}
-                    {{ tojson(exposure.depends_on.nodes) }} {# depends_on_nodes #}
+                    '{{ tojson(exposure.depends_on.nodes) }}', {# depends_on_nodes #}
+                    '{{ tojson(exposure.meta) | replace("'","\\'") }}' {# meta #}
```
Comment (NiallRees) [inline macros/upload_exposures.sql]:
```suggestion
                    parse_json('{{ tojson(exposure.meta) | replace("'","\\'") }}') {# meta #}
```

As this is a key: value object, I think it needs `parse_json`

**Comment by NiallRees** [comment]
Thank you for this @robertholmes3! I've left a few comments but it's 90% there.

**Comment by robertholmes3** [comment]
@NiallRees thanks for the feedback! I made all of the suggested changes. I *Think* I may have brought in some other commit (whoops). Let me know if I need to make any other changes here :)

**Comment by k-lawecki** [comment]
@NiallRees is there anything that should be fixed here that stops this from being merged?
We would love to use meta in exposures :)

**Comment by NiallRees** [comment]
> @NiallRees is there anything that should be fixed here that stops this from being merged? We would love to use meta in exposures :)

Hi @k-lawecki, I'm no longer a member of the Brooklyn Data team but I'm sure someone will be able to help! cc @jecolvin

**Review by jecolvin — APPROVED** [review]

**Comment by k-lawecki** [comment]
@jecolvin are there any plans for release that could include this one? :) We are trying to build an automated exposures catalogue and this would be a massive help

File: models/sources/sources.sql [hunk 6]
```diff
@@ -10,14 +10,17 @@ select
     cast(null as {{ type_string() }}) as database,
     cast(null as {{ type_string() }}) as schema,
     cast(null as {{ type_string() }}) as source_name,
+    cast(null as {{ type_string() }}) as source_description,
```
Comment (glsdown) [inline models/sources/sources.sql:13]:
```suggestion
```

File: models/sources/sources.sql [hunk 7]
```diff
@@ -10,14 +10,17 @@ select
...
-    {% if target.type == 'snowflake'%}
-        cast(null as {{ type_array() }}) as freshness
+    {% if target.type == 'snowflake' %}
+        cast(null as {{ type_array() }}) as freshness,
     {% else %}
-        cast(null as {{ type_json() }}) as freshness
+        cast(null as {{ type_json() }}) as freshness,
     {% endif %}
+        cast(null as {{ type_json() }}) as source_meta,
+        cast(null as {{ type_json() }}) as meta
```
Comment (glsdown) [inline models/sources/sources.sql:24]:
```suggestion
    cast(null as {{ type_string() }}) as source_description,
    cast(null as {{ type_json() }}) as source_meta,
    cast(null as {{ type_json() }}) as meta
```

File: macros/upload_sources.sql [hunk 8]
```diff
@@ -31,11 +34,14 @@
                 '{{ source.database }}', {# database #}
                 '{{ source.schema }}', {# schema #}
                 '{{ source.source_name }}', {# source_name #}
+                '{{ source.source_description | replace("'","\\'") }}', {# source_description #}
```
Comment (glsdown) [inline macros/upload_sources.sql:37]:
```suggestion
```

File: macros/upload_sources.sql [hunk 9]
```diff
@@ -31,11 +34,14 @@
...
                 '{{ source.loader }}', {# loader #}
                 '{{ source.name }}', {# name #}
                 '{{ source.identifier }}', {# identifier #}
                 '{{ source.loaded_at_field | replace("'","\\'") }}', {# loaded_at_field #}
-                '{{ tojson(source.freshness) | replace("'","\\'") }}' {# freshness #}
+                '{{ tojson(source.freshness) | replace("'","\\'") }}', {# freshness #}
+                '{{ tojson(source.source_meta) | replace("'","\\'") }}', {# source_meta #}
+                '{{ tojson(source.meta) | replace("'","\\'") }}' {# meta #}
```
Comment (glsdown) [inline macros/upload_sources.sql:44]:
```suggestion
                '{{ source.source_description | replace("'","\\'") }}', {# source_description #}
                '{{ tojson(source.source_meta) | replace("'","\\'") }}', {# source_meta #}
                '{{ tojson(source.meta) | replace("'","\\'") }}' {# meta #}
```

File: macros/upload_sources.sql [hunk 10]
```diff
@@ -21,7 +21,10 @@
             {{ adapter.dispatch('column_identifier', 'dbt_artifacts')(8) }},
             {{ adapter.dispatch('column_identifier', 'dbt_artifacts')(9) }},
             {{ adapter.dispatch('column_identifier', 'dbt_artifacts')(10) }},
-            {{ adapter.dispatch('parse_json', 'dbt_artifacts')(adapter.dispatch('column_identifier', 'dbt_artifacts')(11)) }}
+            {{ adapter.dispatch('column_identifier', 'dbt_artifacts')(11) }},
+            {{ adapter.dispatch('parse_json', 'dbt_artifacts')(adapter.dispatch('column_identifier', 'dbt_artifacts')(12)) }},
```
Comment (glsdown) [inline macros/upload_sources.sql:25]:
```suggestion
            {{ adapter.dispatch('parse_json', 'dbt_artifacts')(adapter.dispatch('column_identifier', 'dbt_artifacts')(11)) }},
            {{ adapter.dispatch('column_identifier', 'dbt_artifacts')(12) }},
```

File: macros/upload_sources.sql [hunk 11]
```diff
@@ -57,11 +63,14 @@
                     '{{ source.database }}', {# database #}
                     '{{ source.schema }}', {# schema #}
                     '{{ source.source_name }}', {# source_name #}
+                    '{{ source.source_description }}', {# source_description #}
```
Comment (glsdown) [inline macros/upload_sources.sql:66]:
```suggestion
```

File: macros/upload_sources.sql [hunk 12]
```diff
@@ -57,11 +63,14 @@
...
                     '{{ source.loader }}', {# loader #}
                     '{{ source.name }}', {# name #}
                     '{{ source.identifier }}', {# identifier #}
                     '{{ source.loaded_at_field | replace("'","\\'") }}', {# loaded_at_field #}
-                    parse_json('{{ tojson(source.freshness) | replace("'","\\'") }}')  {# freshness #}
+                    parse_json('{{ tojson(source.freshness) | replace("'","\\'") }}'),  {# freshness #}
+                    parse_json('{{ tojson(source.source_meta) | replace("'","\\'") }}'),  {# source_meta #}
```
Comment (glsdown) [inline macros/upload_sources.sql:72]:
```suggestion
                    '{{ source.source_description }}', {# source_description #}
                    parse_json('{{ tojson(source.source_meta) | replace("'","\\'") }}'),  {# source_meta #}
```

**Review by glsdown — CHANGES_REQUESTED** [review]
Hi @robertholmes3 . Sorry that this is so delayed, but we are putting in some time to work through all the open PRs and Issues now. I took a look at this one to get it merged, but was receiving this error:

```
Database Error
  002023 (22000): SQL compilation error:
  Expression type does not match column data type, expecting ARRAY but got VARCHAR(39) for column FRESHNESS
```

This is because when adding columns to the `upload_*` macros, and corresponding `sources` table, they need to be included at the end of the table to avoid clashes with data types and naming. I noticed that `source_description` was added in the middle which seems to have been the cause for it.

Due to the delays getting back to you, I've added comments where the changes would need to be made - only the `macros/upload_sources.sql` and `models/sources/sources.sql` files are affected by this. Please do check these though as it's harder to make the changes in comments than the files themselves.

Sorry again for the delay here.

**Pull request closed without merging**